### PR TITLE
Api CLI

### PR DESCRIPTION
**Pull request by uanabitarte**
Api CLI developed so far: commands for client's operations

### CODE REVIEW

File: .buildpath [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<buildpath>
+	<buildpathentry kind="src" path=""/>
+	<buildpathentry kind="con" path="org.eclipse.php.core.LANGUAGE"/>
+</buildpath>
```
Comment (elacunza) [inline .buildpath]: This file doesn't belong to the repo

File: .project [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+		<buildCommand>
+			<name>org.eclipse.dltk.core.scriptbuilder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
+	</buildSpec>
+	<natures>
+		<nature>org.eclipse.php.core.PHPNature</nature>
+	</natures>
+</projectDescription>
```
Comment (elacunza) [inline .project]: We don't upload project files to the repo

File: .settings/org.eclipse.wst.validation.prefs [hunk 3]
```diff
@@ -0,0 +1,2 @@
+disabled=06vendor
+eclipse.preferences.version=1
```
Comment (elacunza) [inline .settings/org.eclipse.wst.validation.prefs]: No editor settings

File: bin/elkarbackup-api-cli [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+$application = new Application('elkarbackup-api-cli', '@package_version@');
+$application->add(new Command\DeleteClientCommand());
+$application->add(new Command\GetClientsCommand());
+$application->add(new Command\GetClientCommand());
+$application->add(new Command\PostClientCommand());
+$application->add(new Command\PostClientFromFileCommand());
+$application->add(new Command\UpdateClientCommand());
+$application->add(new Command\UpdateClientFromFileCommand());
+$application->run();
```
Comment (elacunza) [inline bin/elkarbackup-api-cli]: Last row with content should have a newline

File: box.json [hunk 5]
```diff
@@ -0,0 +1,4 @@
+{
+    "main": "bin/elkarbackup-api-cli", 
+    "output": "elkarbackup-api-cli.phar"
+}
```
Comment (elacunza) [inline box.json]: Last row with content should have a newline

File: makepackage.sh [hunk 6]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+vendor/bin/box compile
```
Comment (elacunza) [inline makepackage.sh]: Last row with content should have a newline

File: public/index.php [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+if ($trustedHosts = $_SERVER['TRUSTED_HOSTS'] ?? false) {
+    Request::setTrustedHosts([$trustedHosts]);
+}
+
+$kernel = new Kernel($_SERVER['APP_ENV'], (bool) $_SERVER['APP_DEBUG']);
+$request = Request::createFromGlobals();
+$response = $kernel->handle($request);
+$response->send();
+$kernel->terminate($request, $response);
```
Comment (elacunza) [inline public/index.php]: Do we really need a public/index.php file for a CLI project?

File: src/Command/DeleteClientCommand.php [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    }
+    
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $httpClient = HttpClient::create();
+        $username = $input->getArgument('username');
+        $password = $input->getArgument('password');
+        $id = $input->getArgument('id');
+        $output->writeln("Delete client ".$id);
+        $response = $httpClient->request('DELETE', 'http://127.0.0.1/api/clients/'.$id, ['auth_basic' => [$username, $password],]);
```
Comment (elacunza) [inline src/Command/DeleteClientCommand.php]: API URL should be configurable, with a localhost default. We should be able to use CLI command from a remote host. (applies to other commands that have an embedded URL)

**Review by elacunza — CHANGES_REQUESTED** [review]
Some picky changes and others quite important, like embedded API URL.

File: src/Command/DeleteClientCommand.php (hunk 8 above)
Comment (uanabitarte) [reply src/Command/DeleteClientCommand.php]: I have added a parameter to configure the url, but at the moment only works for bin/console. I am having trouble to find a way to make dependency injection work for the symfony console application

File: src/Command/GetClientsCommand.php [hunk 9]
```diff
@@ -19,7 +24,7 @@ protected function configure()
         ->addArgument('password', InputArgument::REQUIRED, "Password for authentication")
         ->addOption('name', null, InputOption::VALUE_REQUIRED, "Filter client list by name")
         ->addArgument('file', InputArgument::OPTIONAL, "Output file for the clients' list")
-        ;
+        ;
```
Comment (elacunza) [inline src/Command/GetClientsCommand.php]: Bogus whitespace change

File: src/Command/DeleteClientCommand.php (hunk 8 above)
Comment (elacunza) [reply src/Command/DeleteClientCommand.php]: This should be a parameter like user and password, I don't understand this dependency injection thing...

Comment (uanabitarte) [reply src/Command/DeleteClientCommand.php]: ok, I was thinking as a configuration variable. I will make the change

File: makepackage.sh (hunk 6 above)
Comment (elacunza) [inline makepackage.sh]: I don't like how this has the same name as "makepackage.sh" in elkarbackup directory, for building a .deb package. Maybe we can call the file "makephar.sh" or just "build.sh" ?

File: src/Command/DeleteClientCommand.php [hunk 10]
```diff
@@ -0,0 +1,38 @@
+<?php
+namespace App\Command;
+
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\HttpClient\HttpClient;
+use Symfony\Component\Console\Input\InputArgument;
+
+class DeleteClientCommand extends Command
```
Comment (elacunza) [inline src/Command/DeleteClientCommand.php]: Why does extend Command and not BaseCommand?

File: src/Command/DeleteClientCommand.php [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+{
+    protected function configure()
+    {
+        $this
+        ->setName('client:delete')
+        ->setDescription('Delete a client')
+        ->addArgument('username', InputArgument::REQUIRED, "Username for authentication")
+        ->addArgument('password', InputArgument::REQUIRED, "Password for authentication")
+        ->addArgument('id', InputArgument::REQUIRED, "Client's id")
+        ->addArgument('url', InputArgument::OPTIONAL, "Url of the api", "http://127.0.0.1");
```
Comment (elacunza) [inline src/Command/DeleteClientCommand.php]: Is there a reason to concatenate all those calls instead of having a line/sentence for each one?
Would it make sense to define a configure() in BaseCommand, with common arguments line url, user and pass, and in children clases' configure() first call parent::configure?

**Comment by elacunza** [comment]
@uanabitarte can you move job CLI to another branch please? We have unresolved things here with client CLI implementation...

**Comment by uanabitarte** [comment]
I have made another PR for the job part #2 . However the last commits (81577f5 ,  01c400f ,  142d374) of this branch are corrections for both client and job commands. Sorry for the little mess

**Comment by elacunza** [comment]
Ok, revert job-related commits and let's sort out last bits on this PR

File: src/Command/BaseCommand.php [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    protected function configure()
+    {
+        $this
+        ->addArgument('username', InputArgument::REQUIRED, "Username for authentication")
+        ->addArgument('password', InputArgument::REQUIRED, "Password for authentication")
+        ->addOption('apiUrl', null, InputOption::VALUE_OPTIONAL, "Url of the api", "http://127.0.0.1");
+        return $this;
+    }
```
Comment (elacunza) [inline src/Command/BaseCommand.php]: Please don't modify parent class method signature. Command::configure returns void .

I don't like this "fluent interface" method call chaining, but if you use it, at least use proper indentation as is shown in Symfony docs:
https://symfony.com/doc/4.4/console.html#configuring-the-command

Note the object line, method call indentation and final ; position

**Review by elacunza — CHANGES_REQUESTED** [review]
Comments for BaseCommand apply to all derived clases.

File: src/Command/BaseCommand.php (hunk 12 above)
Comment (uanabitarte) [reply src/Command/BaseCommand.php]: Ok, but then I cannot call to parent configure function for common arguments. Another way would be to add such arguments on the definition of the application: https://gist.github.com/dhrrgn/8847309 
But then the arguments would get defined outside the command classes. Or we could just repeat the code in each command.

Comment (elacunza) [reply src/Command/BaseCommand.php]: I don't understand. You're modifying an object state:

parent::configure();
$this->addWhatever();

should work, did you try?

Comment (uanabitarte) [reply src/Command/BaseCommand.php]: Yes, it says it is not compatible:

> PHP Fatal error:  Declaration of App\Command\DeleteClientCommand::configure() must be compatible with App\Command\BaseCommand::configure(): void in /home/uanabitarte/elkarbackup_api_cli/src/Command/DeleteClientCommand.php on line 9

File: src/Command/BaseCommand.php [hunk 13]
```diff
@@ -20,13 +20,12 @@ protected function checkRequiredOptionsAreNotEmpty(InputInterface $input): void
         }
     }
 
-    protected function configure()
+    protected function configure(): void
     {
         $this
         ->addArgument('username', InputArgument::REQUIRED, "Username for authentication")
         ->addArgument('password', InputArgument::REQUIRED, "Password for authentication")
         ->addOption('apiUrl', null, InputOption::VALUE_OPTIONAL, "Url of the api", "http://127.0.0.1");
```
Comment (elacunza) [inline src/Command/BaseCommand.php]: Indentation must be fixed

**Review by elacunza — CHANGES_REQUESTED** [review]

File: src/Command/GetClientCommand.php [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $httpClient = HttpClient::create();
+        $url = $input->getOption('apiUrl');
+        $username = $input->getArgument('username');
+        $password = $input->getArgument('password');
+        $id = $this->parseInt($input->getArgument('id'));
+        $response = $httpClient->request('GET', $url.'/api/clients/'.$id, ['auth_basic' => [$username, $password],]);
+        $output->writeln("Get client ".$id);
```
Comment (elacunza) [inline src/Command/GetClientCommand.php]: Don't print "Get client"; just print/write json result if sucessful and no more; that way output can be concatenated/treated like a json.

File: src/Command/GetClientsCommand.php [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+        $name = $input->getOption('name');
+        if ($name) {
+            $filter = "?name=".$name;
+        } else {
+            $filter = null;
+        }
+        $response = $httpClient->request('GET', $url.'/api/clients'.$filter, [
+            'auth_basic' => [$username, $password],
+        ]);
+        $output->writeln("Get clients");
```
Comment (elacunza) [inline src/Command/GetClientsCommand.php]: Same here, Don't print "Get clients"; just print/write json and no more; that way output can be concatenated/treated like a json.

File: src/Command/DeleteClientCommand.php [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+        $password = $input->getArgument('password');
+        $id = $this->parseInt($input->getArgument('id'));
+        $output->writeln("Delete client ".$id);
+        $response = $httpClient->request('DELETE', $url.'/api/clients/'.$id, ['auth_basic' => [$username, $password],]);
+        if (204 == $response->getStatusCode()){
+            $output->writeln("Client ".$id." successfully deleted");
+        } else {
+            $output->writeln($response->getContent());
+        }
+    }
```
Comment (elacunza) [inline src/Command/DeleteClientCommand.php:46]: Command should return 0 if successful, a non-zero value if error. This must be done for all commands. You can check return value of a command typing "echo $?" after the command in a shell.

File: src/Command/GetClientCommand.php [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+        ;
+    }
+    
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $httpClient = HttpClient::create();
+        $url = $input->getOption('apiUrl');
+        $username = $input->getArgument('username');
+        $password = $input->getArgument('password');
+        $id = $this->parseInt($input->getArgument('id'));
```
Comment (elacunza) [inline src/Command/GetClientCommand.php]: This will throw an uncontrolled exception if it is not an int.
This is a probable mistake by the user, not an exception, we should catch it, output an error message and return a suitable error value.

**Review by elacunza — CHANGES_REQUESTED** [review]
Now that first level is sorted, I see more issues. We're arriving to the goal, don't hesitate!

File: src/Command/GetClientCommand.php (hunk 14 above)
Comment (uanabitarte) [reply src/Command/GetClientCommand.php]: Should I remove also the messages like "Client created successfully"?

Comment (elacunza) [reply src/Command/GetClientCommand.php]: In no-GET commands, I think we should maintain the result message and don't print APIs result JSON instead, as user hasn't requested any info. Can issue a subsequent getClient or getClients

**Comment by uanabitarte** [comment]
As right now the API response is only returned in GET and PUT operations, should I remove the output file option for all the commands? I do not think it makes much sense...

**Comment by elacunza** [comment]
Yes, for GET and PUT we can redirect to a file now.

File: src/Command/BaseCommand.php [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+        if (is_numeric($string)) {
+            return (int) $string;
+        }
+        throw new \InvalidArgumentException("Parameter must be integer");
+    }
+
+    protected function returnCode(ResponseInterface $response, OutputInterface $output, $filename = null): int
+    {
+        $message = "";
+        try{
```
Comment (elacunza) [inline src/Command/BaseCommand.php]: missing whitespace

File: src/Command/BaseCommand.php [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+            $status  = $response->getStatusCode();
+            $content = $response->getContent();
+        } catch (\Exception $e) {
+            $message = $e->getMessage();
+        }
+
```
Comment (elacunza) [inline src/Command/BaseCommand.php]: $status doesn't have a value if we caught an exception...

Comment (elacunza) [reply src/Command/BaseCommand.php]: Maybe you're assuming exception will be raised only by getContent?

File: src/Command/BaseCommand.php [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+                if ($filename) {
+                    $file = fopen($filename, 'w');
+                    fwrite($file, $content);
+                    fclose($file);
+                } else {
+                    $output->writeln($content);
+                }
+                return self::SUCCESS;
+            case 201:
+                $output->writeln("Successfully created");
```
Comment (elacunza) [inline src/Command/BaseCommand.php]: I think we should display the ID of the created object

File: src/Command/BaseCommand.php [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+
+    protected function parseInt(string $string): int
+    {
+        if (is_numeric($string)) {
+            return (int) $string;
+        }
+        throw new \InvalidArgumentException("Parameter must be integer");
+    }
+
+    protected function returnCode(ResponseInterface $response, OutputInterface $output, $filename = null): int
```
Comment (elacunza) [inline src/Command/BaseCommand.php]: I don't think it is a good idea to manage child class logic in parent.
You could manage the common errors (auth errors, etc.), but you shouldn't put print logic por 201, 204, etc. they only have sense in a children class context I think.
Also, this method is doing much more that "returnCode"

**Review by elacunza — CHANGES_REQUESTED** [review]
I don't really like the current returnCode method in BaseCommand. It makes things obscure and code hard to follow.

File: src/Command/BaseCommand.php (hunk 19 above)
Comment (uanabitarte) [reply src/Command/BaseCommand.php]: yes, I thought the exception is raised by getContent

Comment (elacunza) [reply src/Command/BaseCommand.php]: https://github.com/symfony/symfony/blob/5.4/src/Symfony/Contracts/HttpClient/ResponseInterface.php

getStatusCode can also raise exceptions. I think you're assuming too much. First get the code and work with it, then ask for the content if needed?

File: src/Command/DeleteClientCommand.php [hunk 22]
```diff
@@ -31,6 +32,16 @@ protected function execute(InputInterface $input, OutputInterface $output): int
             return self::INVALID_ARGUMENT;
         }
         $response = $httpClient->request('DELETE', $url.'/api/clients/'.$id, ['auth_basic' => [$username, $password],]);
-        return $this->returnCode($response, $output);
+        try {
+            $status = $response->getStatusCode();
+        } catch (TransportException $e) {
+            $output->writeln($e->getMessage());
+            return self::ERROR;
```
Comment (elacunza) [inline src/Command/DeleteClientCommand.php]: We should use an error code specific for transport exception (communication error)

File: src/Command/BaseCommand.php [hunk 23]
```diff
@@ -71,9 +72,6 @@ protected function manageError(ResponseInterface $response, OutputInterface $out
         } catch (\Exception $e) {
             $message = $e->getMessage();
             switch ($status) {
-                case 401:
-                    $output->writeln($message);
-                    return self::UNAUTHORIZED;
                 case 404:
```
Comment (elacunza) [inline src/Command/BaseCommand.php:78]: Please undo this modification done for a test
